File: sources/ruby/11_parallel_sum.rb
# task 11 parallel_sum — expected output: 7500000075000000
# build: none (interpreted)    run: ruby 11_parallel_sum.rb (cruby) | ruby --yjit 11_parallel_sum.rb (cruby+yjit) | jruby 11_parallel_sum.rb (jruby, needs Java 25)
# The stock Windows CRuby build has no YJIT: `ruby --yjit` warns "Ruby was built without YJIT support".
# CRuby's GVL serializes the four threads, so this prints the right answer without running any faster;
# JRuby's threads are real JVM threads and do run in parallel.

def work(t)
  acc = 0
  i = t * 25_000_000
  limit = (t + 1) * 25_000_000
  while i < limit
    case i % 4
    when 0 then acc += 1
    when 1 then acc += i
    when 2 then acc += 2 * i
    when 3 then acc += 3 * i
    end
    i += 1
  end
  acc
end

threads = (0...4).map { |t| Thread.new { work(t) } }

total = 0
threads.each { |th| total += th.value }

puts total

File: sources/perl/11_parallel_sum.pl
# task 11 parallel_sum — expected output: 7500000075000000
# build: perl 11_parallel_sum.pl    run: perl 11_parallel_sum.pl
# note: uses threads, so it needs a Perl built with ithreads (standard on Linux and on Strawberry
#       Perl). Each thread runs in its own interpreter, so the partial sums come back via join().
use strict;
use warnings;
use threads;

sub work {
    my ($t) = @_;

    my $acc = 0;
    my $start = $t * 25000000;
    my $limit = ($t + 1) * 25000000;

    for (my $i = $start; $i < $limit; $i++) {
        my $c = $i % 4;
        if ($c == 0) {
            $acc += 1;
        } elsif ($c == 1) {
            $acc += $i;
        } elsif ($c == 2) {
            $acc += 2 * $i;
        } else {
            $acc += 3 * $i;
        }
    }

    return $acc;
}

my @threads = map { threads->create(\&work, $_) } 0 .. 3;

my $total = 0;
$total += $_->join() for @threads;

print $total, "\n";
